File: go/dp_two_dim/322_coin_change.go
package dptwodim

// coinChange 零钱兑换
// 返回能够兑换的最小硬币个数
// 完全背包的视角
func coinChange(coins []int, amount int) int {
	// dp[j] 表示能够凑到amount 所需的最小硬币个数
	// coins[i] 表示物品和物品的价值
	// amount 是背包的容量
	// dp[j] = min(dp[j], dp[j-nums[i]] + 1)

	dp := make([]int, amount+1)
	// 初始化
	// 因为求最小值，初始给个较大值 amount + 1
	dp[0] = 0
	// 考虑到状态转移方程的特性，初始化要是一个大的值
	for i := 1; i <= amount; i++ {
		dp[i] = amount + 1
	}

	for i := 0; i < len(coins); i++ {
		for j := coins[i]; j < amount+1; j++ {
			dp[j] = min(dp[j], dp[j-coins[i]]+1)
		}
	}

	// 不能凑够
	if dp[amount] == amount+1 {
		return -1
	}
	return dp[amount]
}

File: rust/src/dp_two_dim/_322_coin_change.rs
/// 零钱兑换
/// 返回能够兑换的最小硬币个数
/// 完全背包的视角再写一遍
pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
    // dp[j] 表示容量为j的背包填满，最少需要dp[j]个硬币
    // dp[j] = min(dp[j], dp[j-coins[i]] + 1)

    // 最大的硬币数是amout，因为硬币的面值最小是1
    let max_val = amount + 1;

    let amount = amount as usize;
    let mut dp = vec![max_val; amount + 1];
    dp[0] = 0;

    // 遍历物品
    for i in 0..coins.len() {
        // 遍历背包容量
        for j in 0..=amount {
            if j as i32 - coins[i] >= 0 {
                dp[j] = dp[j].min(dp[j - coins[i] as usize] + 1);
            }
        }
    }

    if dp[amount] == max_val {
        return -1;
    }
    dp[amount]
}
